src/analysis/convex/body.lean

```diff
@@ -0,0 +1,126 @@
+
+/--
+Let `V` be a normed space. A subset of `V` is a convex body if and only if
+it is convex, compact, and nonempty.
+-/
+structure convex_body :=
+(carrier : set V)
+(convex' : convex ℝ carrier)
+(is_compact' : is_compact carrier)
+(nonempty' : carrier.nonempty)
+
+namespace convex_body
+
+variables {V}
+
+instance : set_like (convex_body V) V :=
+{ coe := convex_body.carrier,
+  coe_injective' := λ K L h, by { cases K, cases L, congr' } }
+
+lemma convex (K : convex_body V) : convex ℝ (K : set V) := K.convex'
+lemma is_compact (K : convex_body V) : is_compact (K : set V) := K.is_compact'
+lemma nonempty (K : convex_body V) : (K : set V).nonempty := K.nonempty'
+
+@[ext]
+protected lemma ext {K L : convex_body V} (h : (K : set V) = L) : K = L := set_like.ext' h
+
+@[simp]
+lemma coe_mk (s : set V) (h₁ h₂ h₃) : (mk s h₁ h₂ h₃ : set V) = s := rfl
+
+instance : add_monoid (convex_body V) :=
+-- we cannot write K + L to avoid reducibility issues with the set.has_add instance
+{ add := λ K L, ⟨set.image2 (+) K L,
+                 K.convex.add L.convex,
+                 K.is_compact.add L.is_compact,
+                 K.nonempty.add L.nonempty⟩,
+  add_assoc := λ K L M, by { ext, simp only [coe_mk, set.image2_add, add_assoc] },
+  zero := ⟨0, convex_singleton 0, is_compact_singleton, set.singleton_nonempty 0⟩,
+  zero_add := λ K, by { ext, simp only [coe_mk, set.image2_add, zero_add] },
+  add_zero := λ K, by { ext, simp only [coe_mk, set.image2_add, add_zero] } }
+
+@[simp]
+lemma coe_add (K L : convex_body V) : (↑(K + L) : set V) = (K : set V) + L := rfl
+
+@[simp]
+lemma coe_zero : (↑(0 : convex_body V) : set V) = 0 := rfl
+
+instance : inhabited (convex_body V) := ⟨0⟩
+
+instance : add_comm_monoid (convex_body V) :=
+{ add_comm := λ K L, by { ext, simp only [coe_add, add_comm] },
+  .. convex_body.add_monoid }
+
+noncomputable example : has_smul ℝ V := infer_instance
+example : topological_space V := infer_instance
+example : has_continuous_const_smul ℝ≥0 V := @nnreal.has_continuous_const_smul V _ _ _
+
+instance : has_smul ℝ (convex_body V) :=
+{ smul := λ c K, ⟨c • (K : set V), K.convex.smul _, K.is_compact.smul _, K.nonempty.smul_set⟩ }
+
+@[simp]
+lemma coe_smul (c : ℝ) (K : convex_body V) : (↑(c • K) : set V) = c • (K : set V) := rfl
+
+-- @[simp]
+-- lemma coe_smul' (c: ℝ≥0) (K : convex_body V) : (↑(c • K) : set V) = c • (K : set V) := sorry
+
+instance : distrib_mul_action ℝ (convex_body V) :=
+{ to_has_smul := convex_body.has_smul,
+  one_smul := λ K, by { ext, simp only [coe_smul, one_smul] },
+  mul_smul := λ c d K, by { ext, simp only [coe_smul, mul_smul] },
+  smul_add := λ c K L, by { ext, simp only [coe_smul, coe_add, smul_add] },
+  smul_zero := λ c, by { ext, simp only [coe_smul, coe_zero, smul_zero] } }
+
+instance has_smul' : has_smul ℝ≥0 (convex_body V) :=
+nnreal.distrib_mul_action.to_has_smul
+
+@[simp]
+lemma coe_smul' (c : ℝ) (K : convex_body V) : (↑(c • K) : set V) = c • (K : set V) := rfl
```

Comment:
Thanks for adding the ℝ instance! I guess removing this duplicate lemma would fix the linter error, and I guess the `ℝ≥0` `coe_smul'` lemma could be solved by simp so doesn't need to be added?
```suggestion
instance : distrib_mul_action ℝ (convex_body V) :=
{ to_has_smul := convex_body.has_smul,
  one_smul := λ K, by { ext, simp only [coe_smul, one_smul] },
  mul_smul := λ c d K, by { ext, simp only [coe_smul, mul_smul] },
  smul_add := λ c K L, by { ext, simp only [coe_smul, coe_add, smul_add] },
  smul_zero := λ c, by { ext, simp only [coe_smul, coe_zero, smul_zero] } }

instance has_smul' : has_smul ℝ≥0 (convex_body V) :=
nnreal.distrib_mul_action.to_has_smul
```

Reply:
It was not a duplicate lemma, it was supposed to be the `nnreal` equivalent, which I fixed now.